Persistence - Remove Unowned Keys

singku opened the pull request:
Proposing this PR per issue discussion:
https://github.com/valkey-io/valkey/issues/539

For cluster, after startup loading, remove keys
that shouldn't be served by this server based on slot assignment of a cluster.

Also added stat fields in server to count the total removed keys and skipped slots from last loading.

madolson commented on src/server.c:
```diff
@@ -5582,7 +5584,9 @@ sds genValkeyInfoString(dict *section_dict, int all_sections, int everything) {
             "module_fork_in_progress:%d\r\n", server.child_type == CHILD_TYPE_MODULE,
-            "module_fork_last_cow_size:%zu\r\n", server.stat_module_cow_bytes));
+            "module_fork_last_cow_size:%zu\r\n", server.stat_module_cow_bytes,
+            "total_startup_load_deleted_keys:%lld\r\n", server.stat_persistence_startup_load_keys_deleted,
+            "total_startup_load_unowned_slots:%lld\r\n", server.stat_persistence_startup_load_unowned_slots));
```
Let's just log these, I don't think we need to add a server field for these, since they will be static for the lifetime of the instance.

madolson commented on src/cluster_legacy.c:
```diff
@@ -6552,3 +6552,30 @@ void clusterReplicateOpenSlots(void) {
+        return result;
+    }
+
+    clusterNode *master;
```
If we're not a primary (master) we should not be taking any actions here. We should let the primary do the deletions and the replica should follow it. The replica needs to be a carbon copy of the primary, and there are cases where the replica can think it's primary doesn't own a slot it actually does, and won't clean up the data.

singku replied:
Will replication sync remove orphaned keys?

singku replied:
Looks like replica will flush old data before loading rdb by calling emptyData()

singku replied:
Think it second time, I think do it for both replica and primary is a safer way. Consider replica doesn't remove the keys, then before sync with primary, it is promoted to be new primary. Then these data will not be removed.

madolson commented on src/server.c:
```diff
@@ -6941,6 +6945,9 @@ int main(int argc, char **argv) {
         serverLog(LL_NOTICE, "Server initialized");
         aofLoadManifestFromDisk();
         loadDataFromDisk();
+        del_stat stat = delUnOwnedKeys();
```
We should do this later, after "verifyClusterConfigWithData". We want to give this node a chance at taking ownership of this slot.

madolson commented on src/cluster_legacy.c:
```diff
@@ -6552,3 +6552,30 @@ void clusterReplicateOpenSlots(void) {
     zfree(argv);
 }
+
+del_stat delUnOwnedKeys(void) {
```
This should ideally be in cluster.c. The difference is that cluster_legacy is for low level interface, whereas I think we can just use the high level APIs (see cluster.h).

singku replied:
We need to call delKeysInSlot, which is a static/local function defined in this file. Also clusterState is not exposed to cluster.c currently, we need include "cluster_legacy.h" to reference it in cluster.c. Shall we change that then call it from cluster.c?

madolson reviewed (changes requested):
We also should implement a test for this as well. I know it will probably be a bit tricky. I'm thinking of:
1. Create an RDB file with data in all slots.
2. Start a node with it that has a nodes.conf that doesn't own all the slots.

singku commented:
> We also should implement a test for this as well. I know it will probably be a bit tricky. I'm thinking of:
> 
> 1. Create an RDB file with data in all slots.
> 2. Start a node with it that has a nodes.conf that doesn't own all the slots.

If we move the deletion after `verifyClusterConfigWithData` we can't simply do this test because the unassigned slot with keys will be reclaimed by the node. We can only test it in cluster mode with multiple shards.

singku commented:
@madolson gently ping for additional input

hpatro commented on src/cluster.c:
```diff
@@ -107,6 +108,33 @@ ConnectionType *connTypeOfCluster(void) {
+            }
+        }
+    }
+    serverLog(LL_NOTICE, "Deleted totoal: %d unowned keys from total: %d unowned slots after loading RDB",
```
```suggestion
    serverLog(LL_NOTICE, "Deleted %d unowned keys from %d unowned slots after loading RDB", deleted_keys, unowned_slots);
```

hpatro commented on src/cluster.c:
```diff
@@ -107,6 +108,33 @@ ConnectionType *connTypeOfCluster(void) {
+    if (!server.cluster_enabled) {
+        return;
+    }
+    unsigned int deleted_keys = 0;
```
Is `unsigned int` sufficient to store total key count?

singku replied:
outdated code but yes, the existing function uses unsigned int

hpatro commented on src/cluster.c:
```diff
@@ -107,6 +108,33 @@ ConnectionType *connTypeOfCluster(void) {
+    }
+    unsigned int deleted_keys = 0;
+    unsigned int unowned_slots = 0;
+    clusterNode *master;
```
Recently we have moved to `primary/replica` terminology.

```suggestion
    clusterNode *primary;
```

hpatro commented on src/cluster.c:
```diff
@@ -107,6 +108,33 @@ ConnectionType *connTypeOfCluster(void) {
+    unsigned int deleted_keys = 0;
+    unsigned int unowned_slots = 0;
+    clusterNode *master;
+    if (clusterNodeIsMaster(server.cluster->myself)) {
+        master = server.cluster->myself;
+    } else {
+        master = server.cluster->myself->slaveof;
+    }
```
```suggestion    
    primary = clusterNodeGetPrimary(server.cluster->myself);
```

hpatro commented on src/server.h:
```diff
@@ -3535,6 +3535,9 @@ unsigned long LFUDecrAndReturn(robj *o);
 void startEvictionTimeProc(void);
 
+/* delete unowned keys from database at server start up after loading persistence files. */
+void delUnOwnedKeys(void);
```
```suggestion
void delUnownedKeys(void);
```

hpatro commented on src/cluster.c:
```diff
@@ -107,6 +108,33 @@ ConnectionType *connTypeOfCluster(void) {
+    }
+    serverLog(LL_NOTICE, "Deleted totoal: %d unowned keys from total: %d unowned slots after loading RDB",
+              deleted_keys, unowned_slots);
+    return;
```
I don't think we need it.

singku replied:
removed

singku committed:
Persistence - Remove Unowned Keys

For cluster, after startup loading, remove keys
that shouldn't be served by this server based on slot
assignment of a cluster.

Signed-off-by: Liang Tang <tangliang@google.com>

singku commented:
@hpatro Thanks for reviewing this PR, I just rebased it to resolve conflict and applied changes per your comments, please take a look.

hpatro commented on src/cluster.c:
```diff
@@ -108,6 +108,27 @@ ConnectionType *connTypeOfCluster(void) {
+        return;
+    }
+    clusterNode *primary;
+    if (clusterNodeIsPrimary(server.cluster->myself)) {
+        primary = server.cluster->myself;
+    } else {
+        primary = server.cluster->myself->replicaof;
+    }
```
There is a method to fetch the primary i.e. `clusterNodeGetPrimary`, we could use that here.

hpatro commented on src/server.h:
```diff
@@ -3678,6 +3678,9 @@ unsigned long LFUDecrAndReturn(robj *o);
 int performEvictions(void);
 void startEvictionTimeProc(void);
 
+/* delete unowned keys from database at server start up after loading persistence files. */
```
In Valkey project, we generally put the code comment in `.c` file rather than the `.h` file.

hpatro commented on src/cluster.c:
```diff
@@ -108,6 +108,27 @@ ConnectionType *connTypeOfCluster(void) {
+        if (server.cluster->slots[i] != primary) {
+            unsigned int deleted = delKeysInSlot(i);
+            if (deleted > 0) {
+                serverLog(LL_NOTICE, "Deleted %u keys from unowned slot: %d after loading RDB", deleted, i);
```
Can we make it `LL_VERBOSE`, otherwise in the worst case, this might dump lot's of log line (one per missing slot (16384)).

hpatro commented on src/cluster.c:
```diff
@@ -108,6 +108,27 @@ ConnectionType *connTypeOfCluster(void) {
+            }
+        }
+    }
+    return;
```
Do we need the return statement?

hpatro commented:
Mostly LGTM, thanks @singku!

hpatro commented on src/cluster.c:
```diff
@@ -108,6 +108,22 @@ ConnectionType *connTypeOfCluster(void) {
     return connectionTypeTcp();
 }
 
+/* delete unowned keys from database at server start up after loading persistence files. */
```
nit: We can write the first letter in uppercase.

```suggestion 
/* Delete unowned keys from database at server start up after loading persistence files. */
```

hpatro commented:
@madolson Could you take a pass? LGTM.

singku commented:
> @madolson Could you take a pass? LGTM.

Gentle Ping..

PingXie committed:
Update src/cluster.c

Co-authored-by: Harkrishn Patro <bunty.hari@gmail.com>
Signed-off-by: Ping Xie <pingxie@outlook.com>

PingXie commented:
For a more generic solution, I think we need to shed unowned keys *during* loading. The current fix drops the unowned keys *after* the loading is complete but the node might have already OOM'd before reaching the point. 


> We also should implement a test for this as well. 

+1

> If we move the deletion after `verifyClusterConfigWithData` we can't simply do this test because the unassigned slot with keys will be reclaimed by the node. We can only test it in cluster mode with multiple shards, as well performing slot migration in the test

This change does need to be tested in cluster mode. Have you tried out @madolson's [recommendation](https://github.com/valkey-io/valkey/pull/568#pullrequestreview-2083943601)?
